core/src/main/java/org/apache/iceberg/rest/auth/AuthSession.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.auth;
+
+import org.apache.iceberg.rest.HTTPRequest;
+
+/**
+ * An authentication session that can be used to authenticate outgoing HTTP requests.
+ *
+ * <p>Authentication sessions are usually immutable, but may hold resources that need to be released
+ * when the session is no longer needed. Implementations should override {@link #close()} to release
+ * any resources.
+ */
+public interface AuthSession extends AutoCloseable {
```

Comment:
Is `Authenticator` OK for you?

Reply:
Hmm that would look weird imho:

```java
SupportsAuthentication contextualSession = authManager.contextualSession(context, catalogAuth);
SupportsAuthentication tableSession = authManager.tableSession(ident, tableConf, contextualSession);
RESTClient tableClient = client.withSupportsAuthentication(tableSession);
```

I think at least `Authenticator` has the advantage of being a fairly common name, and since the only relevant method in this interface is called `authenticate`, the name sounds reasonable to me.

Reply:
I think the only name in the above example where things get weird is in `client.withSupportsAuthentication(tableSession)`. In that case what about naming this method simply `withAuth(...)` and we could still keep `SupportsAuthentication` as a name for the interface itself

Reply:
Do you mind if we do the rename in an immediate follow-up PR? I am a bit concerned that as soon as I rename to `SupportsAuthentication`, someone else would chime in and suggest something different. Naming is always a bit controversial :-)

Reply:
I would prefer to do the rename here. @danielcweeks do you have any suggestions in terms of a better name for this interface so that we don't clash with the `AuthSession` in `OAuth2Util`?